back/scripts/datasets/topic_aggregator.py

```diff
@@ -0,0 +1,332 @@
+
+    def _normalize_frame(self, df: pd.DataFrame, file: Tuple):
+        """
+        Set of steps to transform a raw DataFrame into a normalized one.
+        """
+        df = (
+            df.pipe(normalize_column_names)
+            .pipe(merge_duplicate_columns)
+            .pipe(safe_rename, self.manual_column_rename)
+            .rename(
+                columns=self.official_topic_schema.set_index("lower_name")["name"].to_dict()
+            )
+            .pipe(self._flag_columns_by_keyword)
+            .pipe(normalize_identifiant, "idBeneficiaire")
+            .pipe(normalize_identifiant, "idAttribuant")
+            .pipe(normalize_montant, "montant")
+        )
+        self._flag_inversion_siret(df, file)
+        self._flag_extra_columns(df, file)
+        return df.pipe(self._select_official_columns).pipe(self._add_metadata, file)
+
+    def _add_metadata(self, df: pd.DataFrame, file: Tuple):
+        """
+        Add to the normalized dataframe infos about the source of the raw file.
+        """
+        optional_features = {}
+        if "idAttribuant" not in df.columns:
+            optional_features["idAttribuant"] = str(file.siren).zfill(9) + "0" * 5
+        return df.assign(
+            topic=self.topic, url=file.url, coll_type=file.type, **optional_features
+        )
+
+    def _select_official_columns(self, frame: pd.DataFrame) -> pd.DataFrame:
+        """
+        Select from the dataframe the columns that are in the official schema.
+        """
+        columns = [x for x in self.official_topic_schema["name"] if x in frame.columns]
+        return frame[columns]
+
+    def _flag_inversion_siret(self, df: pd.DataFrame, file: Tuple):
+        """
+        Flag datasets which have more unique attribuant sire    t than beneficiaire
+        """
+        if "idAttribuant" not in df.columns or "idBeneficiaire" not in df.columns:
+            return
+        n_attribuant = (
+            df["idAttribuant"].str[:9].nunique() if "idAttribuant" in df.columns else 0
+        )
+        n_beneficiaire = df["idBeneficiaire"].str[:9].nunique()
+        if n_attribuant < n_beneficiaire:
+            return
+        LOGGER.error(f"Data with more unique attribuant siret than beneficiaire : {file.url}")
+        raise RuntimeError("Data with more unique attribuant siret than beneficiaire")
+
+    def _flag_columns_by_keyword(self, frame: pd.DataFrame) -> pd.DataFrame:
+        """
+        Identify columns that may correspond to the one in the official schema based on regex.
+        Ensure that this process may not create a duplicate of a column.
+        """
+        extra_colums = set(frame.columns) - set(self.official_topic_schema["name"])
+        matching = {}
+        for col in extra_colums:
+            options = {
+                v
+                for k, v in TOPIC_COLUMNS_NORMALIZATION_REGEX.items()
+                if re.search(k, col.lower()) and v not in frame.columns
```

Comment:
Si on utilise des regex dans des boucles c'est souvent intéressant en terme de performance de les compiler avec un `re.compile`.
Note : ces regex compilées peuvent être utilisées directement comme clé de dictionnaire.

Certaines regex sont peu contraignantes (comme `"nature"`). À voir si c'est plus pertinent d'utiliser `re.fullmatch` à la place de `re.search` qui serait plus contraignant.

Reply:
Pour le point 2, je ne cherche pas exactement nature mais le mot nature dans le nom de la colonne. Je veux matcher "nature de la convention" par exemple.

Reply:
Je suis d'accord pour l'apport de performance.
J'ai ajouté un petit bout de code pour compiler mais je trouve ca assez moche comme méthode. Aurais tu une reco?

Reply:
Pas vraiment, je pense que l'on doit passer par cette étape.
Attention par contre car ça change la façon d'utiliser la regex `if k.search(col.lower()) and v not in frame.columns`

Reply:
Corrigé.